tools/shared.py

```diff
@@ -99,6 +99,51 @@ def run_process(cmd, check=True, input=None, *args, **kw):
   return ret
 
 
+def get_num_cores():
+  import multiprocessing
+  return int(os.environ.get('EMCC_CORES', multiprocessing.cpu_count()))
+
+
+def run_multiple_processes(commands, child_env=None, route_stdout_to_temp_files_suffix=None, pipe_stdout=False, check=True):
+  std_outs = []
+
+  # TODO: Experiment with registering a signal handler here to see if that helps with Ctrl-C locking up the command prompt
+  # when multiple child processes have been spawned.
+  # import signal
+  # def signal_handler(sig, frame):
+  #   sys.exit(1)
+  # signal.signal(signal.SIGINT, signal_handler)
+
+  with ToolchainProfiler.profile_block('run_multiple_processes'):
+    processes = []
+    start = 0
+    end = 0
+    num_parallel_processes = get_num_cores()
+    temp_files = configuration.get_temp_files()
+    while start < len(commands):
+      if start + num_parallel_processes > end and end < len(commands): # Spawn a new process?
+        std_out = temp_files.get(route_stdout_to_temp_files_suffix) if route_stdout_to_temp_files_suffix else (subprocess.PIPE if pipe_stdout else None)
+        if DEBUG:
+          logger.debug('Running subprocess %d/%d: %s' % (end + 1, len(commands), ' '.join(commands[end])))
+        processes += [subprocess.Popen(commands[end], stdout=std_out, env=child_env if child_env else os.environ.copy())]
+        if route_stdout_to_temp_files_suffix:
+          std_outs += [std_out.name]
+        end += 1
+      else:
+        # Too many commands running in parallel, wait for one to finish.
+        out, err = processes[start].communicate()
```

Comment:
> The worst case is that this blocks on a long-running process while all the other cores are idle, I think? I assume that is something `multiprocessing` avoids.

It may be possible that we won't have as tight utilization as a 100% OS-driven select() would have if runtimes are lopsided, but that does not slow this PR down from winning in benchmarks.

Running `embuilder build ALL` in this PR on Windows PC 1 finishes in 

`main`: 342 seconds
`this PR`: 307.4 seconds (+11.3% faster)

On Windows PC 2:

`main`: DNF, python `multiprocessing` hangs due to #13785
`this PR`: 152 seconds (+∞ faster)

In #13465 another use of this PR was benchmarked to be +58.7% faster to use this code instead of Python `multiprocessing`.

> Also, `multiprocessing` keeps up a pool instead of spawning new processes all the time - is that not more efficient as well?

Superficially it might seem like that.. but the reality is that our use of those pooled subprocesses is to just spawn other command line tasks, so we certainly won't gain any performance from having dedicated pooled up python subprocesses for such purpose. They'll run as fast as those cmdline tasks do.

The performance that we seemingly gain from pooling with `multiprocessing` is to avoid/fix performance slowdowns that bringing up python multiprocessing pool itself has. But better yet, if we don't use `multiprocessing` at all, we win 100% of that time that its pooling aims to fix.

I.e. before we'd have (main python process) -> (multiprocessing.pool subprocess) -> (Popen cmdline).

Now we have just (main python process) -> (Popen cmdline). If anything, we win in performance by cutting out the useless middle man python interpreter, which takes time to spin up and interprocess communicate with.

However, to reiterate, this PR does not exist to optimize performance, but to fix correctness. **Even if `multiprocessing` was any amount faster than this, it does not help when it is broken.** (#8013, #718, #13785, etc.)